### PR TITLE
SMP commands for notifications (NKEY/NID and NSUB/NMSG) with separate queue ID and key

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: src/Simplex/Messaging/Client.hs [hunk 1]
```diff
@@ -277,12 +279,27 @@ subscribeSMPQueue c@SMPClient {smpServer, msgQ} rpKey rId =
...
+subscribeSMPQueueNotifications :: SMPClient -> NotifierPrivateKey -> NotifierId -> ExceptT SMPClientError IO ()
+subscribeSMPQueueNotifications = okSMPCommand $ Cmd SNotifier NSUB
+
 -- | Secure the SMP queue by adding a sender public key.
 --
 -- https://github.com/simplex-chat/simplexmq/blob/master/protocol/simplex-messaging.md#secure-queue-command
 secureSMPQueue :: SMPClient -> RecipientPrivateKey -> RecipientId -> SenderPublicKey -> ExceptT SMPClientError IO ()
 secureSMPQueue c rpKey rId senderKey = okSMPCommand (Cmd SRecipient $ KEY senderKey) c rpKey rId
 
+-- | Secure the SMP queue by adding a sender public key.
```
Comment (efim-poberezkin) [inline src/Simplex/Messaging/Client.hs:294]: copy pasted comment

File: src/Simplex/Messaging/Server.hs [hunk 2]
```diff
@@ -72,20 +73,38 @@ runSMPServerBlocking started cfg@ServerConfig {transports} = do
     smpServer :: (MonadUnliftIO m', MonadReader Env m') => m' ()
     smpServer = do
       s <- asks server
-      raceAny_ (serverThread s : map runServer transports)
+      raceAny_
+        ( serverThread s subscribedQ subscribers subscriptions cancelSub :
+          serverThread s ntfSubscribedQ notifiers ntfSubscriptions (\_ -> pure ()) :
```
Comment (efim-poberezkin) [inline src/Simplex/Messaging/Server.hs:78]: will there be commands for notifier to unsub? (e.g. for disabling push notifications)

File: src/Simplex/Messaging/Server.hs [hunk 3]
```diff
@@ -72,20 +73,38 @@ runSMPServerBlocking started cfg@ServerConfig {transports} = do
...
+        updateSubscribers :: STM (Maybe s)
+        updateSubscribers = do
+          (qId, clnt) <- readTBQueue $ subQ s
+          cs <- readTVar $ subs s
+          writeTVar (subs s) $ M.insert qId clnt cs
+          join <$> mapM (notifySubscribers qId) (M.lookup qId cs)
+        notifySubscribers :: QueueId -> Client -> STM (Maybe s)
+        notifySubscribers qId c = do
+          writeTBQueue (rcvQ c) (CorrId B.empty, qId, Cmd SBroker END)
+          stateTVar (clientSubs c) $ \ss -> (M.lookup qId ss, M.delete qId ss)
```
Comment (efim-poberezkin) [inline src/Simplex/Messaging/Server.hs:107]: Somewhat hard to follow what's happening here, maybe too terse.. maybe it's generalizing for both recipients and notifiers.. But making it less terse / adding comments would probably improve readability.

Do I understand correctly that `notifySubscribers` is for ending previous subscriptions for the same queues? If so maybe rename into `notifyPreviousSubscribers` / `endPreviousSubscriptions`, I think it was easier to understand that's what's happening from the previous implementation..

**Review by efim-poberezkin — APPROVED** [review]
please fix comment before merging

File: src/Simplex/Messaging/Server.hs (hunk 2 above)
Comment (epoberezkin) [reply src/Simplex/Messaging/Server.hs:78]: not sure if it is needed, can be added later

File: src/Simplex/Messaging/Server.hs (hunk 3 above)
Comment (epoberezkin) [reply src/Simplex/Messaging/Server.hs:107]: > Do I understand correctly
yes
